src/frontend/query_protocol.ml

```diff
@@ -133,6 +133,27 @@ type occurrences_status =
 
 type occurrence = { loc : Location.t; is_stale : bool }
 
+module Locate_types_result = struct
+  type node_data =
+    | Arrow
+    | Tuple
+    | Object
+    | Poly_variant
+    | Type_ref of
+        { type_ : string;
+          result :
+            [ `Found of string option * Lexing.position
+            | `Builtin of string
+            | `Not_in_env of string
+            | `File_not_found of string
+            | `Not_found of string * string option ]
+        }
+
+  type type_tree = { data : node_data; children : type_tree list }
+
+  type t = Success of type_tree | Invalid_context
+end
```

Comment:
At first I was a bit sad to see this type repeated in `Locate_type`, and thought it would be better to parametrize the tree. But after making the change in I am not sure if it's actually "better". WDYT ? (and @xvw ?)

My attempt: https://github.com/voodoos/merlin/commit/d7cb643668fc4c3f888833e39c51096b0ce99c52